File: soln-stack-problems/remove-all-adjacent-duplicates-in-string.java
/*
https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public String removeDuplicates(String S) {
        Stack<Character> stack=new Stack<>();
        
        for(char ch:S.toCharArray()){
            if(stack.isEmpty())
                stack.push(ch);
            else if(stack.peek()==ch){
                while(!stack.isEmpty() && stack.peek()==ch)
                    stack.pop();
            } 
            else
                stack.push(ch);
        }
        StringBuilder str=new StringBuilder();
        while(!stack.isEmpty()){
            str.append(stack.pop());
        }
        
        return str.reverse().toString();
    }
}

File: C++/soln-stack-problems/remove-all-adjacent-duplicates-in-string.cpp
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string
*/

class Solution {
public:
    string removeDuplicates(string s) {
        stack<char> st;
        for(int i=0;i<s.length();i++) //traversing the stack
        {
            if(!st.empty() && st.top()==s[i]) 
                st.pop();
            
            else 
                st.push(s[i]);  
        }
        
        string str ="";
        while(!st.empty())
        {
            str.push_back(st.top());
            st.pop(); 
        }
        // as the stack reverses the order of elements we reverse str to get the final answer
        reverse(str.begin(),str.end()); 
        return str;
    }
};